Fix Bug 1219957 - Move browser detection code from global.js to mozilla-client.js

kyoshino opened the pull request:
Move the browser detecting functions to the new Mozilla Client module and offer each static properties for faster access. This PR also removes GA events added in #3451.

kyoshino commented:
@alexgibson r?

alexgibson commented on media/js/base/mozilla-client.js:
```diff
@@ -123,17 +185,6 @@ if (typeof Mozilla === 'undefined') {
-
-            // Log the info with GA. Remove this once we have sufficient data
-            window.dataLayer = window.dataLayer || [];
-            window.dataLayer.push({
```
It may take a little while for Gareth to get us GA data from the last PR, so to avoid blocking the merge of this change, should we leave this in for now? We can always file a follow up bug once we have enough data.

kyoshino replied:
Okay, will remove the change from this PR.

alexgibson commented on media/js/base/mozilla-client.js:
```diff
@@ -141,7 +192,7 @@ if (typeof Mozilla === 'undefined') {
 
         // If Firefox is old or for Android, call the fallback function immediately because the API is not implemented
-        if (parseFloat(userVer) < 35 || window.isFirefoxMobile()) {
+        if (parseFloat(userVer) < 35 || this._isFirefoxAndroid()) {
```
Nit: just use `Client` instead of `this`?

kyoshino replied:
:+1:

alexgibson commented:
Thanks @kyoshino - I'll dig into testing the specific pages here, but this looks to be some excellent paying off of tech debt :+1:

kyoshino commented on media/js/base/mozilla-client.js:
```diff
@@ -161,6 +212,20 @@ if (typeof Mozilla === 'undefined') {
+    Client.isFirefox = Client._isFirefox();
+    Client.isFirefoxAndroid = Client._isFirefoxAndroid();
+    Client.isFirefoxiOS = Client._isFirefoxiOS();
+    Client.isFirefoxDesktop = !Client.isFirefoxAndroid;
```
Um, this is wrong. Will fix it.

kyoshino commented:
Fixed the `isFirefoxDesktop` detection. Also, changed `isFirefox` to return `true` also for Firefox for iOS. That means some code now requires `!client.isFirefoxiOS` to exclude iOS, but I think it makes sense.

alexgibson commented on media/js/firefox/new.js:
```diff
@@ -80,8 +58,8 @@
                 // if user is on release channel and has latest version, offer refresh button
-                uiTourGetConfiguration('appinfo', function(config) {
-                    if (config.defaultUpdateChannel === 'release' && config.version === latestFirefoxVersionFull) {
+                client.getFirefoxDetails(function(data) {
```
Should we add an `if (client.isFirefoxDesktop)` conditional here?

kyoshino replied:
:+1:

alexgibson commented on media/js/firefox/sync.js:
```diff
@@ -32,15 +33,15 @@
 
     // Variations 1-5 are Firefox
-    if (window.isFirefox()) {
+    if (client.isFirefox) {
```
Looks like there's nothing for Firefox for iOS to fall into here below, should we add `!client.isFirefoxiOS`?

alexgibson replied:
Actually no need, the new `if (client.isFirefoxDesktop)` below should be enough :+1:

alexgibson commented on media/js/firefox/desktop/intro-anim.js:
```diff
@@ -5,9 +5,11 @@
     function cutsTheMustard() {
         // Bug (1083079) fixed but effects Firefox 35
-        if (window.isFirefox() && window.getFirefoxMasterVersion() === 35 && $('html').hasClass('osx')) {
+        if (client.isFirefox && client.FirefoxMajorVersion === 35 && client.platform === 'osx') {
```
Firefox 35 is quite old now, I think we're good to remove this check as the bug has been fixed for a while :+1:

kyoshino replied:
Okay, will remove this.

alexgibson commented:
This is looking great - I've pushed this branch to demo2 for some testing: https://www-demo2.allizom.org/en-US/. 

Will try and take a closer look at this soon, thanks @kyoshino!

alexgibson commented on media/js/firefox/new.js:
```diff
@@ -43,34 +45,10 @@
+    if (client.isFirefox && !client.isFirefoxiOS) {
+        // Detect whether the Firefox is up-to-date in a non-strict way. The minor version and channel are not
+        // considered. This can/should be strict, once the UX especially for ESR users is decided. (Bug 939470)
+        if (client._isFirefoxUpToDate(false)) {
```
I like they way this is setting us up for a simple change in the near future :+1:

alexgibson commented:
The function `_isFirefoxAndroid()` currently returns true for Firefox on iOS: 

`Mozilla.Client._isFirefoxAndroid('Mozilla/5.0 (iPhone; CPU iPhone OS 8_3 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) FxiOS/1.1 Mobile/12F69 Safari/600.1.4');`

This seems to be the only way that `_isFirefoxDesktop` correctly returns `false` for Firefox on iOS. Both will likely need to be updated.

alexgibson commented on media/js/base/mozilla-client.js:
```diff
@@ -27,8 +100,20 @@ if (typeof Mozilla === 'undefined') {
+     * @return {Number} major version number
+     */
+    Client._getFirefoxMajorVersion = function (ua) {
+        return parseFloat(Client._getFirefoxVersion(ua));
```
Technically, won't this return more than just the major version number? e.g. `Won't this return more than the major version number alone? e.g. `Mozilla.Client._getFirefoxMajorVersion("Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:44.1) Gecko/20100101 Firefox/44.1");`

alexgibson replied:
If we're relying on the expectation that the minor version number is never in the UA string for this to work, I'd feel more comfortable if we explicitly made this return only the major version, like the function name suggests.

kyoshino replied:
I was actually thinking old [releases](https://www.mozilla.org/en-US/firefox/releases/) like 3.6 but yes, 38.4.0 should be considered as `38`, not `38.4`... I'll use `parseInt` instead.

alexgibson commented on media/js/plugincheck/check-plugins.js:
```diff
@@ -171,7 +173,7 @@ $(function() {
 
     // only execute the plugincheck code if this is Firefox
-    if (isFirefox() || isLikeFirefox()) {
+    if ((client.isFirefox && !client.isFirefoxiOS) || client.isLikeFirefox) {
```
Could this be simplified to `client.isFirefoxDesktop || client.isLikeFirefox`?

kyoshino replied:
And `client.isFirefoxAndroid` maybe.

alexgibson commented on media/js/plugincheck/check-plugins.js:
```diff
@@ -157,13 +159,13 @@ $(function() {
 
     // show main download button to non Fx traffic
-    if(!isFirefox() && !isLikeFirefox() ) {
+    if((!client.isFirefox && !client.isLikeFirefox) || client.isFirefoxiOS) {
```
Why would we want to show the download button to users already on Firefox for iOS?

kyoshino replied:
So true.

alexgibson commented on tests/unit/spec/base/mozilla-client.js:
```diff
@@ -9,6 +9,172 @@ describe('mozilla-client.js', function() {
+
+    });
+
+    describe('_isFirefoxAndroid', function () {
```
It would be good to add a test for the Firefox iOS user agent here to make sure it returns false.

kyoshino replied:
Okay

alexgibson commented:
This is looking great, just need to make sure we firm up our UA tests a little for any unexpected false positives :+1:

kyoshino commented:
Fixed the remaining issues @alexgibson have pointed out. I have replaced `client.isFirefox && !client.isFirefoxiOS` with `client.isFirefoxDesktop || client.isFirefoxAndroid` for a better readability.

alexgibson commented on tests/unit/spec/base/mozilla-client.js:
```diff
@@ -9,18 +9,209 @@ describe('mozilla-client.js', function() {
+            expect(window.Mozilla.Client._isFirefoxAndroid('Mozilla/5.0 (Android; Tablet; rv:26.0) Gecko/26.0 Firefox/26.0')).toBeTruthy();
+        });
+
+        it('should return true for Firefox OS on Phone', function() {
```
Wait, why are we returning true here for Firefox OS when the function is `_isFirefoxAndroid`?

kyoshino replied:
Aha, bad copy & paste! Will fix it.

alexgibson commented on media/js/base/mozilla-client.js:
```diff
@@ -10,11 +10,84 @@ if (typeof Mozilla === 'undefined') {
+    Client._isFirefoxAndroid = function (ua) {
+        ua = ua || navigator.userAgent;
+
+        return /\sFirefox/.test(ua) && /Mobile|Tablet|Fennec/.test(ua) && !Client._isLikeFirefox(ua);
```
From the tests it looks like this returns true for Firefox OS? Would this be more robust if we we're checking for `android` specifically?

kyoshino replied:
I just also noticed that Firefox for Maemo still returns `true` here. It's no longer `isFirefoxMobile`...

kyoshino commented:
Alright, the detecting functions are now more robust, and the tests are more readable and comprehensive. I have also added `Client.isFirefoxFxOS` for consistency.

alexgibson commented on media/js/base/mozilla-client.js:
```diff
@@ -10,11 +10,96 @@ if (typeof Mozilla === 'undefined') {
+    Client._isFirefoxFxOS = function (ua) {
+        ua = ua || navigator.userAgent;
+
+        return /\sFirefox/.test(ua) && /Mobile|Tablet/.test(ua) && !/Android/.test(ua);
```
If we're adding Firefox OS to the mix, I think we should remain consistent with our current detection method in site.js: https://github.com/mozilla/bedrock/blob/master/media/js/base/site.js#L67-L69

kyoshino replied:
Thanks, added `navigator.platform` check to be consistent with `site.js`.

alexgibson commented:
@kyoshino - since this Pr was opened a few other pages have been created which use the old `global.js` functions. Could you please rebase and update where needed? Then I'll get this back on a demo server. Thanks for your patience!

kyoshino commented:
@alexgibson Rebased and removed the legacy functions :smiley_cat:

alexgibson commented:
I've put this back up on demo2 and done some preliminary testing and so far things look good. I still need to do some further testing on Android & iOS on certain pages.

alexgibson commented:
I don't have a device with Firefox for Android to test with, so I'm not able to verify the browser specific messaging shown on the following pages:

https://www-demo2.allizom.org/en-US/firefox/android/
https://www-demo2.allizom.org/en-US/firefox/hello/
https://www-demo2.allizom.org/en-US/firefox/ios/
https://www-demo2.allizom.org/en-US/firefox/new/
https://www-demo2.allizom.org/en-US/plugincheck/ (does plugincheck even work on Firefox for Android?)
https://www-demo2.allizom.org/en-US/firefox/private-browsing/
https://www-demo2.allizom.org/en-US/firefox/sync/

@kyoshino - do you have a device to help test with here? If not, I can see if someone from WebQA can help to verify these pages.

kyoshino commented:
I'll test those pages with my Android phone and tablet.

kyoshino commented:
LGTM on Nexus 5X and 9. Confirmed the plugincheck page worked even on Android when [Flash Player was manually installed](https://support.mozilla.org/en-US/kb/how-do-i-watch-flash-videos-firefox-android).

alexgibson commented:
Thanks for helping to verify @kyoshino, and thanks again for this PR. Nice work! r+ :rabbit:

kyoshino commented:
Thanks @alexgibson!
